diff --git i/cmake/config/Config.cmake.in w/cmake/config/Config.cmake.in
index bdde0633e..6cf49f119 100644
--- i/cmake/config/Config.cmake.in
+++ w/cmake/config/Config.cmake.in
@@ -24,5 +24,6 @@ find_dependency(tinyxml2 CONFIG REQUIRED)
 find_dependency(unofficial-sqlite3 CONFIG REQUIRED)
 find_dependency(ada CONFIG REQUIRED)
 find_dependency(WebP CONFIG REQUIRED)
+find_dependency(blend2d CONFIG REQUIRED)
 
 include("${CMAKE_CURRENT_LIST_DIR}/cesium-nativeTargets.cmake")
